core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,394 @@
+   *         - Required context fields (entityType, entityName, rootEntityType) are missing
+   *         - Entity type doesn't match rootEntityType when no entity expression is provided
+   *         - Expressions cannot be parsed or referenced entities/attributes don't exist
+   *
+   */
+  def evaluateExpressions(workspaceId: UUID,
+                          expressionEvaluationContext: ExpressionEvaluationContext,
+                          gatherInputsResult: MethodConfigResolver.GatherInputsResult
+  )(implicit executionContext: ExecutionContext): Future[LazyList[SubmissionValidationEntityInputs]] = {
+    // First, verify that necessary entity information is present and consistent
+    val entityType = expressionEvaluationContext.entityType.getOrElse(
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing entityType"))
+    )
+    val entityName = expressionEvaluationContext.entityName.getOrElse(
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing entityName"))
+    )
+
+    val rootEntityTypeOpt = expressionEvaluationContext.rootEntityType
+    if (rootEntityTypeOpt.isEmpty) {
+      Future.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing rootEntityType")))
+    } else if (
+      expressionEvaluationContext.expression.isEmpty &&
+      entityType != rootEntityTypeOpt.get
+    ) {
+      val whatYouGaveUs =
+        if (expressionEvaluationContext.entityType.isDefined)
+          s"an entity of type ${expressionEvaluationContext.entityType.get}"
+        else "no entity"
+      Future.failed(
+        new RawlsExceptionWithErrorReport(
+          errorReport = ErrorReport(
+            StatusCodes.BadRequest,
+            s"Method configuration expects an entity of type ${rootEntityTypeOpt.get}, but you gave us $whatYouGaveUs."
+          )
+        )
+      )
+    } else {
+
+      // Next, parse both the entity expression (if it exists) and the input expressions
+      // To determine which entities to fetch from the database
+      val rootEntityType = rootEntityTypeOpt.get
+
+      val entityLookups: Seq[ExpressionLookup] = expressionEvaluationContext.expression match {
+        case None             => Seq.empty
+        case Some(expression) => parseLookups(expression)
+      }
+
+      // Parse all input expressions and collect their lookups and parsed trees
+      val inputExpressionData = gatherInputsResult.processableInputs.toSeq.map { input =>
+        val terraExpressionParser = AntlrTerraExpressionParser.getParser(input.expression)
+        val visitor = new CompactEvaluateVisitor()
+        val parsedTree = terraExpressionParser.root()
+        val inputLookups: Seq[ExpressionLookup] = visitor.visit(parsedTree)
+        (input, parsedTree, inputLookups)
+      }
+
+      // Gather all ExpressionLookups from all inputs
+      val allLookups = inputExpressionData.flatMap(_._3)
```

Comment:
use a case statement for readability
```suggestion
      val allLookups = inputExpressionData.flatMap{ case (_, _, lookups) => lookups }
```